crates/macros/src/lib.rs

```diff
@@ -332,6 +336,83 @@ where
     })
 }
 
+#[proc_macro_derive(BorshDeserializer)]
+pub fn derive_borsh_deserializer(struct_def: TokenStream) -> TokenStream {
+    derive_borsh_deserializer_inner(struct_def.into()).into()
+}
+
+#[inline]
+fn derive_borsh_deserializer_inner(item_def: TokenStream2) -> TokenStream2 {
+    let mut hasher = sha2::Sha256::new();
+    let (ident, generics) = syn::parse2::<ItemStruct>(item_def.clone())
+        .map(|def| {
+            hasher.update(def.to_token_stream().to_string().as_bytes());
+            (def.ident, def.generics)
+        })
+        .unwrap_or_else(|_| {
+            let def = syn::parse2::<ItemEnum>(item_def).expect(
+                "BorshDeserializer expected to be derived on a struct or enum",
+            );
+            hasher.update(def.to_token_stream().to_string().as_bytes());
+            (def.ident, def.generics)
+        });
+    let type_hash: [u8; 32] = hasher.finalize().into();
+    let hash = syn::ExprArray {
+        attrs: vec![],
+        bracket_token: Default::default(),
+        elems: Punctuated::<_, _>::from_iter(type_hash.into_iter().map(|b| {
+            syn::Expr::Lit(syn::ExprLit {
+                attrs: vec![],
+                lit: syn::Lit::Byte(LitByte::new(
+                    b,
+                    proc_macro2::Span::call_site(),
+                )),
+            })
+        })),
+    };
+
+    if !generics.params.is_empty() {
+        panic!(
+            "Cannot derive BorshDeserializer on a parameterized type. This \
+             can be done manually for concrete instantiations via the \
+             derive_borshdeserializer! macro."
+        );
+    }
+
+    let deserializer_ident =
+        syn::Ident::new(&format!("{}_DESERIALIZER", ident), Span::call_site());
+    quote!(
+        #[cfg(feature = "migrations")]
+        #[namada_migrations::distributed_slice(REGISTER_DESERIALIZERS)]
+        static #deserializer_ident: fn() = || {
+            let mut locked = namada_migrations::TYPE_DESERIALIZERS.lock().unwrap();
+            locked.insert(#hash, |bytes| {
+                #ident::try_from_slice(&bytes).map(|val| format!("{:?}", val)).ok()
+            });
+        };
+
+        impl namada_migrations::TypeHash for #ident {
+            const HASH: [u8; 32] = #hash;
+        }
+    )
+}
+
+#[allow(unused_macros)]
+macro_rules! derive_borshdeserializer {
```

Comment:
if we promote `derive_borshdeserializer!()` to a procmacro, the hashing can also be computed with `sha2` and we can eliminate the `const-sha2` dep I think. moreoever, we'll also be able to hash the struct fields, and `derive_borsh_deserializer_inne` should be able to be implemented (partially) via `derive_borshdeserializer!()`